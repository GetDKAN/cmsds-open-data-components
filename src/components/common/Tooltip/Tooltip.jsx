import React from 'react'
import cx from 'classnames'
import { usePopperTooltip } from 'react-popper-tooltip'
import PropTypes from 'prop-types'
import './Tooltip.scss'

const Tooltip = ({
  children,
  tooltip,
  hideArrow,
  type,
  customClass,
  tooltipClass,
  show,
  position = 'top'
}) => {
  const {
    getArrowProps,
    getTooltipProps,
    setTooltipRef,
    setTriggerRef,
    visible
  } = usePopperTooltip({
    placement: position
  })

  const Trigger = () => (
    <span
      ref={setTriggerRef}
      className={cx('tooltip-trigger', type, customClass)}
    >
      {children}
    </span>
  )

  return (
    show ? (
      <span className='tooltip-wrapper'>
        {visible && (
          <span
            ref={setTooltipRef}
            {...getTooltipProps({
              className: cx('tooltip-container', type, tooltipClass, { invisible: !show })
            })}
          >
            {!hideArrow && (
              <span
                {...getArrowProps({
                  className: 'tooltip-arrow'
                })}
              />
            )}
            {tooltip}
          </span>
        )}
        {Trigger()}
      </span>
    ) : (
      Trigger()
    )
  )
}

Tooltip.defaultProps = {
  delayShow: 500,
  delayHide: 100,
  hideArrow: true
}

Tooltip.propTypes = {
  /**
   * Tooltip trigger content
   */
  children: PropTypes.any,
  /**
   * Tooltip element content
   */
  tooltip: PropTypes.any,
  /**
   * `true` shows the triangle-arrow on the tooltip container.
   * The arrow is permanently hidden in component SCSS
   */
  hideArrow: PropTypes.bool,
  /**
   * Custom class(es) added to tooltip and trigger wrapping elements
   */
  type: PropTypes.string,
  /**
   * Custom class(es) added to the tooltip trigger wrapping element
   */
  customClass: PropTypes.string,
  /**
   * Custom class(es) added to the tooltip wrapping element
   */
  tooltipClass: PropTypes.string,
  /**
   * `true` allows the tooltip to be shown when invoked
   */
  show: PropTypes.bool,
  /**
   * Position of the tooltip
   */
  position: PropTypes.string
}

Tooltip.displayName = 'Tooltip'
export default Tooltip
