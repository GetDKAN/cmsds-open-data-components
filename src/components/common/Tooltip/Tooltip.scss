@use "../../../scss/modules/modules";
@use "../../../scss/modules/colormap";
@use "../../../scss/modules/functions";
@use "../../../scss/modules/variables";

.tooltip-wrapper {
  position: relative;
  display: inline-flex;
  justify-content: center;
}

.tooltip-trigger {

  // Tooltip with icon trigger inline with text
  &.tooltip-info {
    display: inline-block;
    vertical-align: middle;
    padding: 4px;
    margin: 0 4px 2px;
    line-height: 0;
    border-radius: 100%;
    transition: opacity 0.1s linear;

    &:hover {
      transition: opacity 0.1s linear;
      background-color: colormap.$blueberry-10;
    }

    svg.fa-info-circle {
      font-size: initial;

      path {
        fill: colormap.$blueberry;
      }
    }
  }

  // Tooltip for term definitions in static content
  &.definition {
    @include functions.link-underline(colormap.$charcoal);

    &:hover {
      cursor: help;
    }
  }
}

span.tooltip-container {
  background: rgba(0, 0, 0, 0.8);
  border-radius: 2px;
  border: none;
  box-shadow: none;
  color: white;
  padding: 3px 8px 4px;
  transition: opacity 0.1s linear;
  display: inline-block;
  inset: auto auto 100% auto !important;
  transform: none !important;

  &.definition {
    background: white;
    color: colormap.$charcoal;
    padding: 16px 24px;
    border: 1px solid colormap.$charcoal-20;
    box-shadow: variables.$panel-shadow;
    max-width: 384px;

    .tooltip-arrow {
      display: block;
    }
  }

  .tooltip-arrow {
    display: none;
  }

  &.invisible {
    display: none;
  }

  &.resize-tooltip {
    opacity: 1;
    left: 0;
    // Transition In
    transition-property: opacity, left;
    transition-timing-function: linear;
    transition-duration: 0.1s;

    &.hide-resize-tooltip {
      opacity: 0;
      left: -9999px !important;
      // Transition Out
      transition-property: opacity, left;
      transition-timing-function: linear;
      transition-duration: 0s;
    }
  }
}

@media screen and (max-width: 991px) {
  span.tooltip-container {
    &.definition {
      max-width: 256px;
      font-size: 1.4rem;
      line-height: 1.4;
    }
  }
}
