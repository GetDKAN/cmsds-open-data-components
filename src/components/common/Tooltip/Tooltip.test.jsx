import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import Tooltip from './Tooltip'
import userEvent from '@testing-library/user-event'
import { waitFor } from '@testing-library/react'

const componentArgs = {
  tooltip: 'Tooltip',
  show: true,
  hideArrow: false,
  children: (
    <div style={{ display: 'inline-block' }}>
      <span>Hover over me</span>
    </div>
  )
}

const renderComponent = (args = componentArgs) => render(
  <Tooltip {...args} />
)

describe('Tooltip component.', () => {
  it('Matches snapshot.', () => {
    const renderedTooltip = renderer.create(
      <Tooltip {...componentArgs} />
    ).toJSON()

    expect(renderedTooltip).toMatchSnapshot()
  })

  it('Renders tooltip when \'show\' prop is \'true\'.', async () => {
    renderComponent()

    // Ensure trigger exists
    expect(screen.queryByText('Hover over me')).toBeInTheDocument()

    // Trigger tooltip
    await waitFor(() => userEvent.hover(screen.getByText('Hover over me')))

    // Ensure tooltip shows
    expect(screen.queryByText(componentArgs.tooltip)).toBeInTheDocument()
  })

  it('Does not render tooltip when \'show\' prop is \'false\'.', async () => {
    renderComponent({
      ...componentArgs,
      show: false
    })

    // Ensure trigger exists
    expect(screen.queryByText('Hover over me')).toBeInTheDocument()

    // Trigger tooltip
    await waitFor(() => userEvent.hover(screen.getByText('Hover over me')))

    // Ensure tooltip does not show
    expect(screen.queryByText(componentArgs.tooltip)).not.toBeInTheDocument()
  })

  it('Tooltip shows when triggered.', async () => {
    renderComponent()

    await waitFor(() => userEvent.hover(screen.getByText('Hover over me')))

    expect(screen.queryByText(componentArgs.tooltip)).toBeInTheDocument()
  })
})
