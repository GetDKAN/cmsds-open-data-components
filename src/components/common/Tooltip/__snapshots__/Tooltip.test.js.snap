// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component. Matches snapshot. 1`] = `
<span
  className="tooltip-wrapper"
>
  <span
    className="tooltip-trigger"
  >
    <div
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <span>
        Hover over me
      </span>
    </div>
  </span>
</span>
`;
